File: frontend/src/lib/api/modules.ts
import api from "@/lib/api";
import { AppModule, ModuleSubscription } from "@/lib/types";

export const getModules = async (includeInactive = false): Promise<AppModule[]> => {
    const response = await api.get<AppModule[]>("/modules", {
        params: { include_inactive: includeInactive }
    });
    return response.data;
};

export const getMyModules = async (): Promise<AppModule[]> => {
    const response = await api.get<AppModule[]>("/modules/me");
    return response.data;
};

export const createModule = async (data: {
    codeName: string;
    name: string;
    description?: string;
    isActive?: boolean;
    targetScope?: string;
    priceMonthly?: number | null;
}): Promise<AppModule> => {
    const response = await api.post<AppModule>("/modules", data);
    return response.data;
};

export const updateModule = async (moduleId: string, data: {
    name?: string;
    description?: string;
    isActive?: boolean;
    targetScope?: string;
    priceMonthly?: number | null;
}): Promise<AppModule> => {
    const response = await api.put<AppModule>(`/modules/${moduleId}`, data);
    return response.data;
};

export const getModule = async (moduleId: string): Promise<AppModule> => {
    const response = await api.get<AppModule>(`/modules/${moduleId}`);
    return response.data;
};

export const getSubscriptions = async (params?: {
    companyId?: string;
    userId?: string;
    moduleId?: string;
}): Promise<ModuleSubscription[]> => {
    const queryParams: Record<string, string> = {};
    if (params?.companyId) queryParams.company_id = params.companyId;
    if (params?.userId) queryParams.user_id = params.userId;
    if (params?.moduleId) queryParams.module_id = params.moduleId;
    
    const response = await api.get<ModuleSubscription[]>("/modules/subscriptions", { params: queryParams });
    return response.data;
};


export const createSubscription = async (data: {
    moduleId: string;
    companyId?: string;
    userId?: string;
    scope: string;
    status?: string;
    expiresAt?: string;
    notes?: string;
}): Promise<ModuleSubscription> => {
    const response = await api.post<ModuleSubscription>("/modules/subscriptions", data);
    return response.data;
};

export const updateSubscription = async (subId: string, data: {
    status?: string;
    expiresAt?: string | null;
    notes?: string;
}): Promise<ModuleSubscription> => {
    const response = await api.put<ModuleSubscription>(`/modules/subscriptions/${subId}`, data);
    return response.data;
};

export const deleteSubscription = async (subId: string): Promise<void> => {
    await api.delete(`/modules/subscriptions/${subId}`);
};

File: backend/routers/modules.py
"""
Módulos y Suscripciones SaaS.

Gestión del catálogo de módulos de la plataforma y las suscripciones
de empresa o usuario a dichos módulos.
"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session, joinedload
from typing import List, Optional
from uuid import UUID
from datetime import datetime

import crud, models, schemas, auth
from database import get_db

router = APIRouter(prefix="/modules", tags=["modules"])


# ─── Helpers ────────────────────────────────────────────────────────────────

def _require_admin(current_user: models.User):
    if not getattr(current_user, "is_platform_admin", False):
        raise HTTPException(status_code=403, detail="Solo el administrador de la plataforma puede realizar esta acción.")


# ─── Catálogo de Módulos ─────────────────────────────────────────────────────

@router.get("", response_model=List[schemas.AppModuleResponse])
def list_modules(
    include_inactive: bool = False,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """Lista todos los módulos disponibles. Admin puede ver también los inactivos."""
    if include_inactive and not getattr(current_user, "is_platform_admin", False):
        include_inactive = False  # Non-admin cannot see inactive
    return crud.get_modules(db, include_inactive=include_inactive)


@router.post("", response_model=schemas.AppModuleResponse)
def create_module(
    data: schemas.AppModuleCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """Crea un nuevo módulo en el catálogo (Platform Admin)."""
    _require_admin(current_user)

    existing = crud.get_module_by_code(db, data.code_name)
    if existing:
        raise HTTPException(status_code=400, detail=f"Ya existe un módulo con el código '{data.code_name}'.")

    return crud.create_module(db, data)


@router.put("/{module_id}", response_model=schemas.AppModuleResponse)
def update_module(
    module_id: str,
    data: schemas.AppModuleUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """Actualiza un módulo del catálogo (Platform Admin)."""
    _require_admin(current_user)

    updated = crud.update_module(db, module_id, data)
    if not updated:
        raise HTTPException(status_code=404, detail="Módulo no encontrado.")
    return updated


# ─── Suscripciones ──────────────────────────────────────────────────────────

@router.get("/subscriptions", response_model=List[schemas.ModuleSubscriptionResponse])
def list_subscriptions(
    company_id: Optional[str] = None,
    user_id: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """
    Lista suscripciones.
    Platform Admin puede ver todas.
    Manager puede ver las de su empresa activa.
    """
    is_admin = getattr(current_user, "is_platform_admin", False)
    active_cid = str(getattr(current_user, "active_company_id", None) or "")
    active_role = getattr(current_user, "active_role", None)

    if not is_admin:
        # Non-admin can only see subscriptions for their own active company
        if active_role != "manager" or not active_cid:
            raise HTTPException(status_code=403, detail="No autorizado.")
        company_id = active_cid
        user_id = None  # Managers cannot filter by user_id

    subs = db.query(models.ModuleSubscription)\
        .options(joinedload(models.ModuleSubscription.module))\
        .filter(
            (models.ModuleSubscription.company_id == company_id) if company_id else True,
            (models.ModuleSubscription.user_id == user_id) if user_id else True
        ).order_by(models.ModuleSubscription.created_at.desc()).all()

    return subs


@router.post("/subscriptions", response_model=schemas.ModuleSubscriptionResponse)
def create_subscription(
    data: schemas.ModuleSubscriptionCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """Crea una suscripción a un módulo (Platform Admin)."""
    _require_admin(current_user)

    # Verify module exists
    module = crud.get_module_by_id(db, str(data.module_id))
    if not module:
        raise HTTPException(status_code=404, detail="Módulo no encontrado.")

    try:
        sub = crud.create_subscription(db, data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    # Reload with module relation
    db.refresh(sub)
    return sub


@router.put("/subscriptions/{sub_id}", response_model=schemas.ModuleSubscriptionResponse)
def update_subscription(
    sub_id: str,
    data: schemas.ModuleSubscriptionUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """Actualiza el estado/expiración de una suscripción (Platform Admin)."""
    _require_admin(current_user)

    updated = crud.update_subscription(db, sub_id, data)
    if not updated:
        raise HTTPException(status_code=404, detail="Suscripción no encontrada.")
    return updated


@router.delete("/subscriptions/{sub_id}")
def delete_subscription(
    sub_id: str,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """Elimina una suscripción (Platform Admin)."""
    _require_admin(current_user)

    deleted = crud.delete_subscription(db, sub_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Suscripción no encontrada.")
    return {"ok": True}


# ─── Módulos del Usuario Actual ──────────────────────────────────────────────

@router.get("/me", response_model=List[schemas.AppModuleResponse])
def get_my_modules(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(auth.get_verified_user)
):
    """
    Devuelve la lista de módulos activos a los que el usuario actual tiene acceso
    (bien por suscripción personal o por suscripción de su empresa activa).
    Útil para que el frontend sepa qué funciones mostrar.
    """
    if getattr(current_user, "is_platform_admin", False):
        # Platform Admin tiene acceso a todos los módulos activos
        return crud.get_modules(db, include_inactive=False)

    user_id = str(current_user.id)
    company_id = str(getattr(current_user, "active_company_id", None) or "")
    now = datetime.utcnow()
    active_statuses = [models.SubscriptionStatus.active, models.SubscriptionStatus.trial]

    from sqlalchemy import or_

    # Subs activas del usuario o de su empresa
    subs = db.query(models.ModuleSubscription)\
        .options(joinedload(models.ModuleSubscription.module))\
        .filter(
            models.ModuleSubscription.status.in_(active_statuses),
            or_(
                models.ModuleSubscription.expires_at.is_(None),
                models.ModuleSubscription.expires_at > now
            ),
            or_(
                models.ModuleSubscription.user_id == user_id,
                models.ModuleSubscription.company_id == company_id if company_id else False
            )
        ).all()

    # Devolver los módulos únicos (puede haber sub personal + sub empresa del mismo módulo)
    seen = set()
    result = []
    for sub in subs:
        if sub.module and sub.module.is_active and sub.module.id not in seen:
            seen.add(sub.module.id)
            result.append(sub.module)

    return result
